code/game/objects/items/weapons/tools.dm

```diff
@@ -337,48 +337,39 @@
 			if(prob(5))
 				remove_fuel(1)
 
-		//If you're actually actively welding, use fuel faster.
-		//Is this actually used or set anywhere? - Nodrak
-		if(2)
-			if(prob(75))
-				remove_fuel(1)
-
-
-	//I'm not sure what this does. I assume it has to do with starting fires...
-	//...but it doesnt check to see if the welder is on or not.
 	var/turf/location = src.loc
 	if(istype(location, /mob/))
 		var/mob/M = location
 		if(M.is_holding_item(src))
 			location = get_turf(M)
-	if (istype(location, /turf))
+	if (istype(location, /turf) && welding)
 		location.hotspot_expose(source_temperature, 5,surfaces=istype(loc,/turf))
 
 
-/obj/item/weapon/weldingtool/afterattack(obj/O as obj, mob/user as mob, proximity)
+/obj/item/weapon/weldingtool/afterattack(atom/A, mob/user as mob, proximity)
 	if(!proximity)
 		return
-	if (istype(O, /obj/structure/reagent_dispensers/fueltank) && get_dist(src,O) <= 1 && !src.welding)
-		O.reagents.trans_to(src, max_fuel)
-		to_chat(user, "<span class='notice'>Welder refueled</span>")
-		playsound(src, 'sound/effects/refill.ogg', 50, 1, -6)
+	if (istype(A, /obj/structure/reagent_dispensers/fueltank) && get_dist(src,A) <= 1 && !src.welding)
+		if(A.reagents.trans_to(src, max_fuel))
+			to_chat(user, "<span class='notice'>Welder refueled.</span>")
+			playsound(src, 'sound/effects/refill.ogg', 50, 1, -6)
+		else if(!A.reagents)
+			to_chat(user, "<span class='notice'>\The [A] is empty.</span>")
+		else
+			to_chat(user, "<span class='notice'>\The [src] is already full.</span>")
 		return
-	else if (istype(O, /obj/structure/reagent_dispensers/fueltank) && get_dist(src,O) <= 1 && src.welding)
+	else if (istype(A, /obj/structure/reagent_dispensers/fueltank) && get_dist(src,A) <= 1 && src.welding)
```

Comment:
This is not your fault, but really we shouldn't have this be a snowflake behavior of fuel-tanks. It should be, for example, chem heating for fuel (hitting something with a welder heats up its contents) or even just a welder behavior (preattack, if it hits something containing, say, 200u fuel or more, explode).

I say this because just a day or two ago a clown blew up by welding an empty welding tank. At the same time, why doesn't a chemist explode if they slam one of their huge Erlenmeyer flasks full of fuel with a welder?